Adds basic usage section to IPFS. (#1117)
ipfs/ipfs-docs: docs/basics/command-line.md

---
title: "Command-line"
description: "A simple walkthrough of how to perform basic IPFS operations using the command-line."
---

# Basic CLI Operations

> This short guide aims to walk you through the basics of using IPFS with the CLI (Command Line Interface). You will learn how to add, retrieve, read, and remove files from within the CLI. We will try to keep it simple here, if you are unsure about the meaning of some terms you can check out the [glossary](../concepts/glossary/). Also don't forget to check out the rest of the [docs](../) to learn more about IPFS!

Reviewer: ```suggestion
This short guide aims to walk you through the basics of using IPFS with the CLI. You will learn how to add, retrieve, read, and remove files from within the CLI. We will try to keep it simple here. If you are unsure about the meaning of some terms, check out the [glossary](../concepts/glossary/). Also don't forget to check out the rest of the [docs](../) to learn more about IPFS!
```

We can expect users to know what a CLI is.

> All instructions and examples shown here were performed and tested on an M1 Mac, the IPFS commands will be same on macOS, Windows, and Linux. You will need to know how to navigate through your computers directories from within the CLI, if you are unsure how to do that, it is recommended to learn how before continuing with this guide.

Reviewer: ```suggestion
All instructions and examples shown here were performed and tested on an M1 Mac. However, the IPFS commands are the same on Linux, MacOS, and Windows. You will need to know how to navigate through your computers directories from within the CLI, if you are unsure how to do that, it is recommended to learn how before continuing with this guide.
```

Try to keep Linux, MacOS, and Windows in alphabetical order. Stops the haters.

## Install IPFS using the CLI

> Installing IPFS using the CLI is simple, and there is already a great guide on how to do so [here](../install/command-line/).

Reviewer: Don't like to pages using `here`. Use the link to tell me what's on that page:

```
We've got a great guide on [using the CLI to interact with IPFS](../install/command-line).
```

Something like that.

Once you have IPFS installed, you need to get your node up and running. If this is your first time using IPFS you will first need to initialize the configuration files.

> ```

Reviewer: ````suggestion
```shell
````

Always give code blocks a syntax to use. If there isn't a syntax that fits, use `plaintext`.

ipfs init

> initializing ipfs node at /Users/<user>/.ipfs
> generating 2048-bit RSA keypair...done
> peer identity: Qm...
> to get started, enter:
>
>   ipfs cat /ipfs/QmYwAPJzv5CZsnA625s3Xf2nemtYgPpHdWEz79ojWnPbdG/readme
```

Your IPFS node is now initialized and you are ready to start the daemon to bring your node online.

```

> ifps daemon

Reviewer: Typo

> Initializing daemon...
> go-ipfs version: 0.12.0
> Repo version: 12
> System version: arm64/darwin
> ...
```

You're all set! Simply open up a new instance of the CLI and continue on with the guide!

> ### Important Note
> 
> Do not close the CLI that you used to initialize your daemon, only terminate the daemon when you want to take your IPFS node offline.

Reviewer: Add notes like this into callout:

```suggestion
:::warning
Do not close the CLI that you used to initialize your daemon, only terminate the daemon when you want to take your IPFS node offline.
:::
```

This is a VuePress specific thing. You can't use this code on Lotus or Filecoin docs.

> ## Add a local file or folder to IPFS using the CLI

Reviewer: This title is really long. Can we shorten it?

Now that you have your IPFS node up and running, you are ready to add your file or folder to IPFS.

> 1. Within the CLI, navigate to the directory that contains the file/folder that you wish to host and share on IPFS. In this example, I will be navigating to my documents directory to add a single text file.
> 
> ```
> cd Documents
> 
> > Documents
> ```

Reviewer: ````suggestion
1. Within the CLI, navigate to the directory that contains the file/folder that you wish to host and share on IPFS. In this example, I will be navigating to my documents directory to add a single text file.

    ```
    cd Documents

    > Documents
    ```
````

If you've got a codeblock in a list item, indent the codeblock so that it formats properly on the site.

2. Once you are in the directory that contains the file/folder you wish to host and share on IPFS, you will be using the `ipfs add <"File/Folder Name">` command to add it to IPFS.

```
ipfs add "Hello IPFS"

> added QmYPpRuJRsR9jGy8dDRJ3txLRWByoT3CJ3AfmKtMq8CPG3 Hello IPFS

> ```

Reviewer: See above comment.

When you add a local file or folder to IPFS it is automatically pinned to your node. Pins prevent objects on your node from being removed during garbage collection, more on this later.

You have now added your files to IPFS and they are ready to be shared and hosted to peers on the network.

> ## Retrieve a file from IPFS and read it using the CLI

Reviewer: Again, this title is pretty long.

In the previous section we went through how to add local files to IPFS, now we will learn how to retrieve remote files from IPFS and save them to your computer locally. For this example we will be retrieving a folder that contains a single text file.

1. Within the terminal, navigate to the directory where you wish to save the folder. The folder will be saved to whichever directory you are in. In this case, I want to save the folder to my documents directory.

```
cd Documents

> Documents

> ```

Reviewer: Codeblocks, indent. You get the idea.

2. Now that you are in the desired destination directory, use the command `ipfs get <CID>` to retrieve the folder.

```
ipfs get bafybeifsohflj3hk4mond6pla2utxux6xblsevrwfeluc6rjij4pdywtlu

> Saving file(s) to bafybeifsohflj3hk4mond6pla2utxux6xblsevrwfeluc6rjij4pdywtlu
> 1.76 KiB / 1.76 KiB [==============================================] 100.00% 0s

> ```

Reviewer: Indent codeblocks in lists.

You have now retrieved the folder over IPFS and a copy of it has been saved to your computers local storage. You are also now hosting the folder and it's contents for others to retrieve.

You can now pin the folder that you just retrieved by running the `ipfs pin add <CID>` command.

```
ipfs pin add bafybeifsohflj3hk4mond6pla2utxux6xblsevrwfeluc6rjij4pdywtlu

> pinned bafybeifsohflj3hk4mond6pla2utxux6xblsevrwfeluc6rjij4pdywtlu recursively
```

> Objects that you retrieve over IPFS are not pinned to your node by default, if you wish to prevent the files from being garbage collected, you need to pin them. You will notice that the pin you just added is a `recursive` pin, meaning it is a directory containing other objects or is a file consisting of only one block. More about pins [here](../how-to/pin-files/#three-kinds-of-pins).

Reviewer: Don't use `here` as link text. Tell the reader what they're about to click on.
